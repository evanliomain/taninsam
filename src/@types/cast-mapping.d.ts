export interface CastMapping {
  [key: string]: (x: any) => any;
}
