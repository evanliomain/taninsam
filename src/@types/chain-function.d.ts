import { Links } from './links';

export type ChainFunction<T, U> = (input: T, links?: Links) => U;
