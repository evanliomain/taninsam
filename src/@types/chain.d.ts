import { Links } from './links';

export interface Chain<T> {
  value(): T;
  chain<U>(f: (x: T, links: Links) => U): Chain<U>;
  link(linkName: string): Chain<T>;
  log(flag: string): Chain<T>;
}
