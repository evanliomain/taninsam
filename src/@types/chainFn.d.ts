import { Links } from './links';

export interface ChainFn<T> {
  value: () => (e) => T;
  chain<U>(g: (x: T, links?: Links) => U): ChainFn<U>;
  link(linkName: string): ChainFn<T>;
  log(flag: string): ChainFn<T>;
}
