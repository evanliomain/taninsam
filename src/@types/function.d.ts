import { Links } from './links';
export type Function<T, U> = (x: T, links?: Links) => U;
