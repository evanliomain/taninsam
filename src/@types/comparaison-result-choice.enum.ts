/**
 * @ignore
 */
export enum ComparaisonResultChoice {
  Greater = 1,
  Lower = -1,
  Equal = 0,
}
