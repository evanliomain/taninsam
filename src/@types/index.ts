export * from './chain-function';
export * from './chain';
export * from './chainFn';
export * from './comparaison-result-choice.enum';
export * from './function';
export * from './identity';
export * from './iteratee';
export * from './links';
export * from './result';
