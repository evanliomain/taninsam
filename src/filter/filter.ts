/**
 * @module array=>array
 */
import { Identity, Links } from '../@types';

/**
 * This method filter the elements of the chained array.
 * See the [native filter](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/filter) for more informations.
 * @param iteratee The iteratee invoked per element to decide to keep or not the element.
 * @return the function to apply on the array to filter it
 * @example Keep even number from an array
 * ```
 * filter<number>((x: number) => 0 === x % 2)([1, 2, 3, 4, 5]) // [2, 4]
 * ```
 * @example Using the chain
 * ```
 * chain([1, 2, 3, 4, 5])
 *   .chain(filter(x => 0 === x % 2))
 *   .value() // [2, 4]
 * ```
 */
export function filter<T>(
  iteratee: ((
    value: T,
    index: number,
    array: ReadonlyArray<T>,
    links?: Links
  ) => boolean),
  links?: Links
): Identity<ReadonlyArray<T>> {
  return (array: ReadonlyArray<T>) =>
    array.filter((value: T, index: number, array: ReadonlyArray<T>) =>
      iteratee(value, index, array, links)
    );
}
