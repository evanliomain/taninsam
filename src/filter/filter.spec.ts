import { describe, expect, test, vi } from 'vitest';
import { filter } from './filter';
import { Links } from '../@types';

describe('filter function', () => {
  const input = [1, 2, 3, 4, 5];

  test('is a pure function', () => {
    filter<number>((x: number) => 0 === x % 2)(input);
    expect(input).toMatchSnapshot();
  });

  test('keep even [1, 2, 3, 4, 5] |> filter === [2, 4]', () => {
    expect(filter<number>((x: number) => 0 === x % 2)(input)).toMatchSnapshot();
  });

  test('call the iteree callback with correct inputs', () => {
    const cb = vi.fn(() => true);
    const link: Links = { fake: 'fake' };
    filter<number>(cb)(input, link);
    input.forEach((i, index) => {
      expect(cb.mock.calls[index][0]).toBe(i);
      expect(cb.mock.calls[index][1]).toBe(index);
      expect(cb.mock.calls[index][2]).toEqual(input);
      expect(cb.mock.calls[index][3]).toBe(link);
    });
  });
});
