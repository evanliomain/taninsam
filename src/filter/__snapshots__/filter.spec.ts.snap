// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`filter function > is a pure function 1`] = `
[
  1,
  2,
  3,
  4,
  5,
]
`;

exports[`filter function > keep even [1, 2, 3, 4, 5] |> filter === [2, 4] 1`] = `
[
  2,
  4,
]
`;
